File: backend-api/app/models/product.rb
class Product < ApplicationRecord
  include Rails.application.routes.url_helpers
  belongs_to :organization,class_name:"Organization",foreign_key: :org_id
  belongs_to :category

  has_many :variants, class_name: 'ProductVariant', dependent: :destroy
  has_many :product_attributes, class_name: 'ProductAttribute', dependent: :destroy
  has_many :cart_items, dependent: :destroy
  has_many :order_items, dependent: :destroy
  has_many_attached :images

  accepts_nested_attributes_for :variants, allow_destroy: true,reject_if: :all_blank
  accepts_nested_attributes_for :product_attributes, allow_destroy: true,reject_if: :all_blank

  validates :name, presence: true
  validates :slug, presence: true, uniqueness: { scope: :org_id, message: "should be unique within the organization" }
  validates :org_id, presence: true
  validates :category_id, presence: true
  validates :status, inclusion: { in: %w[active inactive archived], message: "%{value} is not a valid status" }

  attr_accessor :remove_image
  before_save :purge_images_if_requested

  validate :must_have_at_least_one_variant

  private

  def purge_images_if_requested
    if ActiveModel::Type::Boolean.new.cast(remove_image)
      images.purge if images.attached?
    end
  end

  def must_have_at_least_one_variant
    remaining = variants.reject(&:marked_for_destruction?)
    if remaining.empty?
      errors.add(:base, "Product must have at least one variant")
    end
  end

end

File: web-app/src/types/product.ts
export interface ProductAttribute {
    id?: number | string;
    key: string;
    value: string;
    _destroy?: boolean;
}

export interface ProductVariant {
    id?: number | string;
    name: string;
    sku: string;
    price: number;
    stock: number;
    _destroy?: boolean;
}

export interface Product {
    id: number | string;
    name: string;
    price:number;
    slug: string;
    description?: string;
    status: 'active' | 'inactive' | 'archived';
    org_id?: number | string;
    category_id: number | string;
    category?: {
        id: number | string;
        name: string;
    };
    variants: ProductVariant[];
    product_attributes: ProductAttribute[];
    images?: string[];
}

export interface ProductFormData extends Omit<Product, 'id' | 'variants' | 'product_attributes' | 'category'> {
    id?: number | string;
    variants_attributes?: ProductVariant[];
    product_attributes_attributes?: ProductAttribute[];
    remove_image?: boolean;
}
